torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -355,6 +355,89 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
   return {std::move(payload), std::move(tensors)};
 }
 
+tensorpipe::Message tensorpipeSerialize(const Message& rpcMessage) {
+  tensorpipe::Message tpMessage;
+  const std::vector<char>& payload = rpcMessage.payload();
+  const std::vector<torch::Tensor>& tensors = rpcMessage.tensors();
+
+  // Payload
+  tpMessage.data = (uint8_t*)(payload.data());
+  tpMessage.length = payload.size();
+
+  // Metadata - encode rpc message type + message id
+  tpMessage.metadata = std::to_string(static_cast<int>(rpcMessage.type())) +
+      ":" + std::to_string(static_cast<int>(rpcMessage.id()));
+
+  // Tensors
+  tpMessage.tensors.reserve(tensors.size());
+  // Iterate with copy of tensor to make refcount of tensorImplright
+  for (at::Tensor tensor : tensors) {
+    tensorpipe::Message::Tensor tpTensor;
+    torch::jit::Pickler pickler(
+        [&](const void* buf, size_t sz) -> size_t {
+          tpTensor.metadata.append(static_cast<const char*>(buf), sz);
+          return sz;
+        },
+        nullptr);
+    pickler.protocol();
+    pickler.pushIValue(tensor);
+    pickler.stop();
+    const auto& tensorDataVec = pickler.tensorData();
+    TORCH_INTERNAL_ASSERT(
+        tensorDataVec.size() == 1, "There should be single pickled tensor");
+    const auto& tensorData = tensorDataVec.front();
+    // NOTE: tensorData shares the same storage with rpcMessage tensor
```

Comment:
1. How is the code guaranteeing the lifetime of the source Tensor?
   I don't see a place in tpMessage where you're persisting the tensors.

2. This copy avoidance on write isn't universally safe, since it's legal to call with torch::from_blob() tensors, where the user can [unwisely] choose to manually manage tensor memory. I fixed a related bug in this space in ProcessGroupAgent earlier this week, probably won't be the last.

To mitigate, ThriftRpcAgent's WireSerializer checks whether the tensor has a deleter (grep for storageHasDeleter()). If there's no deleter, it has to assume that the user is manually managing memory, and has to force a copy.

Reply:
@jjlilley Thanks for the insights!
1. From the discussion with @mrshenli, seems users are supposed to keep the tensor buffer lifetime and content consistency.
2. I see. Will do the same check to take sparse tensor and storageHasDeleter() into considerations.

Reply:
Re: 1- I want to make sure we're on the same page here.

- If the caller calls ``future = agent.send(dest, {tensor})``, and then the tensor goes out of scope, how is TensorPipes agents keeping the tensor alive to get the bits out to the kernel, particularly with the io send on another thread and might suffer arbitrary delays?

- fwiw, both ProcessGroupAgent and ThriftRpcAgent will persist the passed-in Tensor data as long as is necessary to get the bits into the kernel.
   They do it in different ways (and Thrift avoids a copy), but they still keep the Tensor refcount alive across the caller->io_thread transition.
   Neither guarantees what bits will be sent if the caller mutates the tensor immediately after the async call starts, but they do guarantee no read-after-free.

It's possible that I missed the mechanism, but it's going to be fairly essential, even in a minimalist impl, to maintain the lifetime of the shared buffer appropriately.

(The (2) point is fairly minor in comparison)